src/main/java/com/minecolonies/rankup/util/AccountingUtils.java

```diff
@@ -0,0 +1,524 @@
+    {
+        try
+        {
+            if (plugin.getModuleContainer().isModuleLoaded(DatabaseModule.ID))
+            {
+                final ResultSet results = getQuery("SELECT " + JOIN_DATE_COLUMN + " FROM " + table_id + " WHERE " + UUID_COLUMN + " = '" + uuid + "'");
+
+                if (results != null && results.next())
+                {
+                    return results.getDate(JOIN_DATE_COLUMN).toString();
+                }
+            }
+            else
+            {
+                return getAccountConfig().playerData.get(uuid).joinDate;
+            }
+        }
+        catch (SQLException | NoModuleException e)
+        {
+            e.printStackTrace();
+        }
+        return "";
+    }
+
+    public String getPlayerLastDate(final UUID uuid)
+    {
+        try
+        {
+            if (plugin.getModuleContainer().isModuleLoaded(DatabaseModule.ID))
+            {
+                final ResultSet results = getQuery("SELECT " + LAST_JOIN_COLUMN + " FROM " + table_id + " WHERE " + UUID_COLUMN + " = '" + uuid + "'");
+
+                if (results != null && results.next())
+                {
+                    return results.getDate(LAST_JOIN_COLUMN).toString();
+                }
+            }
+            else
+            {
+                return getAccountConfig().playerData.get(uuid).lastVisit;
+            }
+        }
+        catch (SQLException | NoModuleException e)
+        {
+            e.printStackTrace();
+        }
+        return "";
+    }
+
+    public int getPlayerTime(final UUID uuid)
+    {
+        try
+        {
+            if (plugin.getModuleContainer().isModuleLoaded(DatabaseModule.ID))
+            {
+                final ResultSet results = getQuery("SELECT " + TIME_PLAYED_COLUMN + " FROM " + table_id + " WHERE " + UUID_COLUMN + " = '" + uuid + "'");
+
+                if (results != null && results.next())
+                {
+                    return results.getInt(TIME_PLAYED_COLUMN);
+                }
+            }
+            else
+            {
+                return getAccountConfig().playerData.get(uuid).timePlayed;
+            }
+        }
+        catch (SQLException | NoModuleException e)
+        {
+            e.printStackTrace();
+        }
+        return 0;
+    }
+
+    private void updatePlayerAttribute(final UUID uuid, final String column, final String attribute, final boolean isDate) throws SQLException
+    {
+        final Connection conn = getConn();
+
+        if (conn != null)
+        {
+            Statement stmt = conn.createStatement();
+
+            if (isDate)
+            {
+                stmt.execute("UPDATE " + table_id + " SET " + column + " = " + toDate(attribute) + " WHERE " + UUID_COLUMN + " = '" + uuid + "'");
+            }
+            else
+            {
+                stmt.execute("UPDATE " + table_id + " SET " + column + " = '" + attribute + "' WHERE " + UUID_COLUMN + " = '" + uuid + "'");
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use a variable binding mechanism to construct this query instead of concatenation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orionminecraft.com/coding_rules#rule_key=squid%3AS2077)